### PR TITLE
Run DiffusionGemma and LLaDA2 pipeline tests through the shared PipelineTesterMixin

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Follow-up to #14652. Runs `DiffusionGemma` and `LLaDA2` pipeline tests through the shared `PipelineTesterMixin`, the same way `audioldm2` does it for a non-image output, skipping the tests that assume image/video semantics (CFG, `latents`, per-row generator lists). Existing hand-written generation/regression tests stay as-is alongside.

Running the mixin's save/load and to-device tests on a real accelerator caught two real bugs, fixed here too:
- `DiffusionGemma`'s canvas init and all three block-diffusion schedulers called `torch.randint`/`multinomial`/`rand` with a CPU generator on a CUDA tensor, which `torch` rejects. Fixed like `randn_tensor` does: sample on the generator's device, move after.
- `LLaDA2Pipeline.tokenizer` defaults to `None` and works fine without one, but wasn't declared in `_optional_components`, so save/load silently dropped it.

Includes akshan-main's #14652 commit as a base — needed for DiffusionGemma's own save/load tests to pass.

Self-review notes (`/code-review medium`): caught two more `torch.rand` calls in `add_noise` with the same generator/device bug (fixed), and DiffusionGemma's pipeline duplicating the device-selection logic instead of reusing the new shared helper (fixed). Also flagged that the `_optional_components` fix wasn't actually exercised since that mixin test was skipped — adapted it to run with `input_ids` instead.

## Before submitting
- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR?
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)?
- [x] Was this discussed/approved via a GitHub issue or the forum? Discussed on #14652.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@sayakpaul

### CODE REVIEW

File: src/diffusers/pipelines/diffusion_gemma/pipeline_diffusion_gemma.py [hunk 1]
```diff
@@ -344,9 +345,12 @@ def __call__(
             )
 
             # Start from a fully random canvas and denoise it; the scheduler resets its committed state at step 0.
+            # `torch.randint` requires the generator and the output device to match, so (as with `randn_tensor`) a
+            # CPU generator samples on CPU and the result is moved to `device` afterwards.
+            rand_device = _generator_device(cur_input_ids, generator)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/diffusion_gemma/pipeline_diffusion_gemma.py]: Why do we need a utility for this?

File: src/diffusers/schedulers/scheduling_discrete_ddim.py [hunk 2]
```diff
@@ -117,7 +117,12 @@ def _sample_from_logits(
             token = flat_logits.argmax(dim=-1, keepdim=True)
         else:
             scaled_probs = torch.softmax(flat_logits.float() / temperature, dim=-1)
-            token = torch.multinomial(scaled_probs, num_samples=1, generator=generator)
+            # `torch.multinomial` requires the generator and the sampled tensor's device to match; `_generator_device`
+            # gives the CPU-generator portability `randn_tensor` gives the continuous samplers.
+            rand_device = _generator_device(scaled_probs, generator)
+            token = torch.multinomial(scaled_probs.to(rand_device), num_samples=1, generator=generator).to(
+                scaled_probs.device
+            )
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_discrete_ddim.py]: Same.

File: src/diffusers/schedulers/scheduling_utils.py [hunk 3]
```diff
@@ -27,6 +27,15 @@
 SCHEDULER_CONFIG_NAME = "scheduler_config.json"
 
 
+def _generator_device(tensor: torch.Tensor, generator: torch.Generator | None) -> torch.device:
+    """Device to sample on for a discrete (`torch.randint`/`torch.multinomial`/`torch.rand`) draw: the generator's
+    own device when one is passed, else `tensor`'s device. Mirrors `randn_tensor`'s CPU-generator portability (a CPU
+    generator, the recommended default for reproducible pipeline calls, samples on CPU regardless of `tensor`'s device)
+    for the discrete samplers, which have no `randn_tensor` equivalent to call into.
+    """
+    return generator.device if generator is not None else tensor.device
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_utils.py]: It is a one liner. Let's fold this into the caller sites.

**Review by sayakpaul — COMMENTED** [review]
Thanks!

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
@yiyixuxu could you also review the changes introduced in the core files?

File: src/diffusers/schedulers/scheduling_discrete_ddim.py [hunk 4]
```diff
@@ -117,7 +117,12 @@ def _sample_from_logits(
             token = flat_logits.argmax(dim=-1, keepdim=True)
         else:
             scaled_probs = torch.softmax(flat_logits.float() / temperature, dim=-1)
-            token = torch.multinomial(scaled_probs, num_samples=1, generator=generator)
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_discrete_ddim.py:120]: Do we use this scheduler in any of the pipelines for dLLM? If not, let's remove these changes.

Comment (kashif) [reply src/diffusers/schedulers/scheduling_discrete_ddim.py:120]: Yes, DiffusionGemma supports it directly and our tests exercise it, so it's in scope.

**Review by yiyixuxu — APPROVED** [review]
thanks!
